src/computability/regular_expressions.lean

```diff
@@ -0,0 +1,429 @@
+    { right,
+      exact hy } }
+end
+
+instance : semiring (language α) :=
+{ add := λ l m, l + m,
+  add_assoc := by simp [set.union_assoc],
+  zero := 0,
+  zero_add := by simp,
+  add_zero := by simp,
+  add_comm := by simp [set.union_comm],
+  mul := λ l m, l * m,
+  mul_assoc := mul_assoc,
+  zero_mul := by simp,
+  mul_zero := by simp,
+  one := 1,
+  one_mul := one_mul,
+  mul_one := mul_one,
+  left_distrib := left_distrib,
+  right_distrib := right_distrib }
+
+lemma add_self (l : language α) : l + l = l := by finish
+
+end language
+
+
+/-- This is the definition of regular expressions. The names used here is to mirror the definition
+  of a Kleene algebra (https://en.wikipedia.org/wiki/Kleene_algebra).
+  `zero` matches nothing
+  `epsilon` matches only the empty string
+  `char a` matches only the string 'a'
+  `star P` matches any finite concatenation of strings which match `P`
+  `plus P Q` matches anything which match `P` or `Q`
+  `comp P Q` matches `x ++ y` if `x` matches `P` and `y` matches `Q` -/
+inductive regular_expression (α : Type u) : Type u
+| zero : regular_expression
+| epsilon : regular_expression
+| char : α → regular_expression
+| plus : regular_expression → regular_expression → regular_expression
+| comp : regular_expression → regular_expression → regular_expression
+| star : regular_expression → regular_expression
+
+namespace regular_expression
+
+instance : inhabited (regular_expression α) := ⟨zero⟩
+
+instance : has_add (regular_expression α) := ⟨plus⟩
+instance : has_mul (regular_expression α) := ⟨comp⟩
+instance : has_one (regular_expression α) := ⟨epsilon⟩
+instance : has_zero (regular_expression α) := ⟨zero⟩
+
+@[simp] lemma zero_def : (zero : regular_expression α) = 0 := rfl
+@[simp] lemma one_def : (epsilon : regular_expression α) = 1 := rfl
+
+@[simp] lemma plus_def (P Q : regular_expression α) : plus P Q = P + Q := rfl
+@[simp] lemma comp_def (P Q : regular_expression α) : comp P Q = P * Q := rfl
+
+/-- `matches P` provides a language which contains all strings that `P` matches -/
+def matches : regular_expression α → language α
+| zero := 0
+| epsilon := 1
+| (char a) := {[a]}
+| (plus P Q) := P.matches + Q.matches
+| (comp P Q) := P.matches * Q.matches
+| (star P) := P.matches.star
+
+@[simp] lemma matches_zero_def : (0 : regular_expression α).matches = 0 := rfl
+@[simp] lemma matches_epsilon_def : (1 : regular_expression α).matches = 1 := rfl
+@[simp] lemma matches_add_def (P Q : regular_expression α) :
+  (P + Q).matches = P.matches + Q.matches := rfl
+@[simp] lemma matches_mul_def (P Q : regular_expression α) :
+  (P * Q).matches = P.matches * Q.matches := rfl
+@[simp] lemma matches_star_def (P : regular_expression α) : P.star.matches = P.matches.star := rfl
+
+/-- `match_epsilon P` is true if and only if `P` matches the empty string -/
+def match_epsilon : regular_expression α → bool
+| zero := ff
+| epsilon := tt
+| (char _) := ff
+| (plus P Q) := P.match_epsilon || Q.match_epsilon
```

Comment:
```suggestion
| 0 := ff
| 1 := tt
| (char _) := ff
| (P + Q) := P.match_epsilon || Q.match_epsilon
```
When you introduce notation, try to use it consistently.

Reply:
I wasn't aware that notation could be used in pattern matching like this without using `@[pattern]` in some way that I don't know about

Reply:
Weirdly this doesn't seem to be working for `comp`

Reply:
> I wasn't aware that notation could be used in pattern matching like this without using `@[pattern]` in some way that I don't know about

It works for `+` because `has_add.add` has the `@[pattern]` attribute.  It fails for `*` (i.e., comp) because `has_mul.mul` doesn't have that attribute.